tests/storage/export.go

```diff
@@ -1194,14 +1277,116 @@ var _ = SIGDescribe("Export", func() {
+		vm = startVM(vm)
+		waitForExportPhase(export, exportv1.Pending)
+		waitForExportCondition(export, expectedVMRunningCondition(vm.Name, vm.Namespace), "export should report VM running")
+	})
+
+	It("should report export pending if PVC is in use because of VMI using it, and start the VM export if the PVC is not in use, then stop again once pvc in use again", func() {
+		sc, exists := libstorage.GetRWOFileSystemStorageClass()
+		if !exists {
+			Skip("Skip test when Filesystem storage is not present")
+		}
+		dataVolume := libstorage.NewDataVolumeWithRegistryImportInStorageClass(cd.DataVolumeImportUrlForContainerDisk(cd.ContainerDiskAlpine), util.NamespaceTestDefault, sc, k8sv1.ReadWriteOnce, k8sv1.PersistentVolumeFilesystem)
+		dataVolume.Spec.PVC.Resources.Requests[k8sv1.ResourceStorage] = resource.MustParse("6Gi")
+		dataVolume = createDataVolume(dataVolume)
+		vmi := tests.NewRandomVMIWithDataVolume(dataVolume.Name)
+		vmi = createVMI(vmi)
+		Eventually(func() virtv1.VirtualMachineInstancePhase {
+			vmi, err := virtClient.VirtualMachineInstance(vmi.Namespace).Get(vmi.Name, &metav1.GetOptions{})
+			if errors.IsNotFound(err) {
 				return ""
 			}
-			return export.Status.Phase
-		}, 30*time.Second, time.Second).Should(Equal(exportv1.Skipped))
+			Expect(err).ToNot(HaveOccurred())
+			return vmi.Status.Phase
+		}, 180*time.Second, time.Second).Should(Equal(virtv1.Running))
+		// For testing the token is the name of the source VM.
+		token := createExportTokenSecret(vmi.Name, vmi.Namespace)
+		pvcName := ""
+		for _, volume := range vmi.Spec.Volumes {
+			if volume.DataVolume != nil {
+				pvcName = volume.DataVolume.Name
+			}
+		}
+		Expect(pvcName).ToNot(BeEmpty())
+		export := createPVCExportObject(pvcName, vmi.Namespace, token)
+		Expect(export).ToNot(BeNil())
+		waitForExportPhase(export, exportv1.Pending)
+		waitForExportCondition(export, expectedPVCInUseCondition(dataVolume.Name, dataVolume.Namespace), "export should report pvc in use")
+
+		By("Deleting VMI, we should get the export ready eventually")
+		deleteVMI(vmi)
+		export = waitForReadyExport(export)
+		verifyKubevirtInternal(export, export.Name, export.Namespace, vmi.Spec.Volumes[0].DataVolume.Name)
+		By("Starting VMI, the export should return to pending")
+		vmi = tests.NewRandomVMIWithDataVolume(dataVolume.Name)
+		vmi = createVMI(vmi)
+		waitForExportPhase(export, exportv1.Pending)
+		waitForExportCondition(export, expectedPVCInUseCondition(dataVolume.Name, dataVolume.Namespace), "export should report pvc in use")
+	})
+
+	It("should mark the status phase skipped on VM without volumes", func() {
+		vm := tests.NewRandomVMWithEphemeralDisk(cd.ContainerDiskFor(cd.ContainerDiskCirros))
+		vm = createVM(vm)
+		// For testing the token is the name of the source VM.
+		token := createExportTokenSecret(vm.Name, vm.Namespace)
+		export := createVMExportObject(vm.Name, vm.Namespace, token)
+		Expect(export).ToNot(BeNil())
+		waitForExportPhase(export, exportv1.Skipped)
 	})
 })
```

Comment:
I think we need a test for VM with no (initial?) persistent volumes.  How should that be handled?  Pending?  Skipped?

Reply:
Should be skipped based on the unit test. I can add a functional test for it.